### ISSUE TITLE
Feature Request: Add support for port discovery for a service 

### ISSUE THREAD

**Issue by playground** [open-horizon/anax#3821]
### Is your feature request related to a problem? Please describe.

If the agent is registered with two services each happens to be bound to the same port, the second service will fail.  

### Describe the solution you'd like.

It will be good to have a way to detect and resolve port conflicts at the time of registering the agent.  

### Describe alternatives you've considered

Is it possible to be able to perform this task via API or from the console?

**Comment by dianzhu** [comment]
@playground @joewxboy Hi Joe, I am Dian Zhu. This issue interests me. I think we can create a port scan tool that detects if port is opened or closed. Please assign it to me if you think this issue is needed.

**Comment by joewxboy** [comment]
@dlarson04 FYI ... in case you have any recommendations for how this should function.
@dabooz FYI ... in case you have any thoughts or opinions about this.

**Comment by dianzhu** [comment]
@joewxboy @dlarson04 @dabooz Do you think if we need to use AI? Do yo recommend Python or Java code?

**Comment by joewxboy** [comment]
It might be helpful to perform this deconfliction in a few places:
* In a validation step during service publishing to warn when there are micro-services (including dependencies) referenced in a single service definition file that share a port
* In an agreement negotiation phase where the agent (anax) running on a node can be told what resources (including ports) a service will need so that the agent can factor those resources into the agreement logic and not agree to deploying a micro-service due to a potential port conflict or other resource contention and respond accordingly.

**Comment by dianzhu** [comment]
@joewxboy @dlarson04 @dabooz It sounds interesting. If we need to use AI, what type of AI tools we should use?

**Comment by dlarson04** [comment]
@joewxboy @dianzhu

I am not sure how valid this is... A node has no way to know what services it will run at registration time so it doesn't know what ports it needs to have open.  Likewise when publishing a service, the new service will have no idea if it will be deployed to an agent with another service that could have a conflict.  It could be done at agreement phase but the agent I am thinking this would only give the agent the opportunity to reject the proposal and not run the service, but this is basically the same net result as the agent trying to start the 2nd service and failing.... ie. in both cases the 2nd service would not start. 

We do have a feature called in the deployment string called `ephemeral_ports`
https://open-horizon.github.io/docs/anax/docs/deployment_string/

This just says use any available port on the host to map to the port my service is using... Running 2 services with this enabled

```
"ephemeral_ports":[   {"port_and_protocol":"8300/tcp"} ],
```		

Yielded
```
# docker ps
CONTAINER ID   IMAGE                                                COMMAND          CREATED          STATUS          PORTS                     NAMES
0d0fd8b9438d   quay.io/abcd04/hzn-http-service-multiarch:0.0.6   "/hzn-service"   12 minutes ago   Up 12 minutes   0.0.0.0:49154->8300/tcp   b6d3695d7d392a8e3fe137562d9f6b1cb30b373e9dd64d5b481961151c76f3a7-ephemeral1
e5110bc78f4d   quay.io/abcd04/hzn-http-service-multiarch:0.0.6   "/hzn-service"   12 minutes ago   Up 12 minutes   0.0.0.0:49153->8300/tcp   000cce7055f607b3cb743e06730142d53e66e3f15a25121eb3d713e6b14a9fd8-ephemeral2
```

**Comment by dianzhu** [comment]
@joewxboy @dlarson04 It looks that agent can run on 2 services using different ports at same time?

**Comment by dlarson04** [comment]
@dianzhu    Yes it can but but it depends on what is in the service definition... If the service does not 
specify ephemeral ports but just has

```
"port":[{"HostPort":"8300/tcp"}]
```

Then 2 of these will not able to run since the first service will grab the port

**Comment by joewxboy** [comment]
@dianzhu and @playground can we discuss in the Agent Working Group meeting today?

**Comment by dianzhu** [comment]
@joewxboy sorry Joe, I missed message above. Please unassign it to me because I feel it's a little difficult for me

**Comment by joewxboy** [comment]
Done, @dianzhu , thanks.

**Comment by playground** [comment]
Not a priority now... and according to @dlarson04, there is way to make it work now.
